### ISSUE TITLE
ipv6 packet sender

### ISSUE THREAD

**Issue by DaryaTsikach** [libpnet/libpnet#344]
Hello,
I was trying to send ipv6 packet using pnet on transport channel at layer3. It was surprisingly for me, that send_to on layer 3 is not implemented for ipv6.  (I am trying to do this on my kubuntu 18.04 laptop)
I was wondering if I am doing something wrong? And if I am, could you please clarify how pnet is supposed to work with ipv6 packets?

Best regards!

**Comment by mrmonday** [comment]
It is highly platform-dependent as to whether this works. IPv6 support at layer 3 on most operating systems is either not supported, or not well supported.

It looks like it should be possible to get this working on Linux - I'd be willing to mentor you if you would like to work on it.

I would personally recommend working with the datalink layer instead, and manually constructing the Ethernet frame. It has far better platform support, and far better performance.

Sorry I took a while to respond, and sorry I can't be of more use!

**Comment by WGH-** [comment]
> I would personally recommend working with the datalink layer instead, and manually constructing the Ethernet frame. It has far better platform support, and far better performance.

Doing ARP manually is not fun, though :)

**Comment by WGH-** [comment]
I tested this only on Linux.

A few gotchas:
* `IP_HDRINCL` doesn't work (`setsockopt(3, SOL_IP, IP_HDRINCL, [1], 4) = -1 ENOPROTOOPT (Protocol not available)`). It seems the working way to do this is to set proto 255 (`IPPROTO_RAW`), even though some sources claim 255 has no special meaning for raw IPv6 sockets.
* Source address is not filled in automatically.

```diff
diff --git a/pnet_transport/src/lib.rs b/pnet_transport/src/lib.rs
index 1a22417..382a17a 100644
--- a/pnet_transport/src/lib.rs
+++ b/pnet_transport/src/lib.rs
@@ -29,7 +29,7 @@ use pnet_packet::udp::UdpPacket;
 use pnet_packet::icmp::IcmpPacket;
 use pnet_packet::icmpv6::Icmpv6Packet;
 use pnet_packet::tcp::TcpPacket;
-use self::TransportChannelType::{Layer3, Layer4};
+use self::TransportChannelType::{Layer3, Layer4, Layer3Ipv6};
 use self::TransportProtocol::{Ipv4, Ipv6};
 
 use std::io;
@@ -56,6 +56,8 @@ pub enum TransportChannelType {
     Layer4(TransportProtocol),
     /// The application will send and receive IPv4 packets, with the specified transport protocol
     Layer3(IpNextHeaderProtocol),
+    /// The application will send and receive IPv6 packets
+    Layer3Ipv6,
 }
 
 /// Structure used for sending at the transport layer. Should be created with transport_channel()
@@ -110,7 +112,11 @@ pub fn transport_channel(buffer_size: usize,
             Layer4(Ipv6(IpNextHeaderProtocol(proto))) => {
                 pnet_sys::socket(pnet_sys::AF_INET6, pnet_sys::SOCK_RAW, proto as libc::c_int)
             }
+            Layer3Ipv6 => {
+                pnet_sys::socket(pnet_sys::AF_INET6, pnet_sys::SOCK_RAW, libc::IPPROTO_RAW)
+            }
         }
+
     };
     if socket == pnet_sys::INVALID_SOCKET {
         return Err(Error::last_os_error());
```

**Comment by JuxhinDB** [comment]
Thanks for the follow-up @WGH- -- would you like to open a PR regarding the above? I'd gladly help.

**Comment by WGH-** [comment]
Yes, why not. I'll test it on different platforms first. Give me a few weeks :)

I also want to re-check whether protocol supplied in `socket` has any effect when `IP_HDRINCL` is enabled. Maybe `Layer3(IpNextHeaderProtocol)` also needs to be redesigned.

**Comment by JuxhinDB** [comment]
That's fine, take all the time you need. If I don't respond, feel free to ping me directly.

**Comment by infosechoudini** [comment]
Can we close this? #377  was closed due to inactivity. @WGH- @JuxhinDB
